### PR TITLE
Propagate SkuResult errors to JS layer for fetch_order_credential

### PR DESCRIPTION
**Pull request by nvonpentz**
Resolves https://github.com/brave/brave-browser/issues/33973

Looks like this:
image

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by bsclifton — APPROVED** [review]
Changes LGTM!

File: components/skus/browser/skus_service_impl.cc [hunk 1]
```diff
@@ -32,8 +32,15 @@ void OnFetchOrderCredentials(
     skus::FetchOrderCredentialsCallbackState* callback_state,
     skus::SkusResult result) {
   if (callback_state->cb) {
-    std::move(callback_state->cb).Run("");
+    const int error_code = static_cast<int>(result);
+    std::string error_message;
+    if (error_code != 0) {
+      error_message = static_cast<std::string>(skus::result_to_string(result));
```
Comment (rillian) [inline components/skus/browser/skus_service_impl.cc:38]: This follows local style, as is fine, but I was confused by the cast operator here. `cxx::rust::String` implements `operator std::string` but it's not triggered implicitly by the assignment. @bsclifton said the casts were just to select it.

Seems like `message = std::string{rust_string}` would be shorter, or if we're trying to avoid triggering the copy ctor, `static_cast<std::string&&>(...)` would be more appropriate.

**Review by rillian — APPROVED** [review]
rust++

File: components/skus/browser/skus_service_impl.cc [hunk 2]
```diff
@@ -32,8 +32,15 @@ void OnFetchOrderCredentials(
     skus::FetchOrderCredentialsCallbackState* callback_state,
     skus::SkusResult result) {
   if (callback_state->cb) {
-    std::move(callback_state->cb).Run("");
+    const int error_code = static_cast<int>(result);
+    std::string error_message;
+    if (error_code != 0) {
```
Comment (evq) [inline components/skus/browser/skus_service_impl.cc:37]: `skus::SkusResult` should be a C++ enum, I think this would be better as `if (error_code != skus::SkusResult::Ok)`

File: components/skus/renderer/skus_js_handler.cc [hunk 3]
```diff
@@ -194,9 +194,16 @@ void SkusJSHandler::OnFetchOrderCredentials(
                                  v8::MicrotasksScope::kDoNotRunMicrotasks);
 
   v8::Local<v8::Promise::Resolver> resolver = promise_resolver.Get(isolate);
+
+  if (!response.empty()) {
```
Comment (bridiver) [inline components/skus/renderer/skus_js_handler.cc:198]: This is a bit hard to follow at first, I think maybe something like this would be better
```
v8::Local<v8::String> result = v8::String::NewFromUtf8(isolate, response.c_str()).ToLocalChecked();
if (response.empty()) {
  std::ignore = resolver->Resolve(context, result);
} else {
  std::ignore = resolver->Reject(context, result);
}

Comment (bridiver) [reply components/skus/renderer/skus_js_handler.cc:198]: also this needs a test

File: components/skus/browser/skus_service_impl.cc [hunk 4]
```diff
@@ -32,8 +32,14 @@ void OnFetchOrderCredentials(
     skus::FetchOrderCredentialsCallbackState* callback_state,
     skus::SkusResult result) {
   if (callback_state->cb) {
-    std::move(callback_state->cb).Run("");
+    std::string error_message;
+    if (result != skus::SkusResult::Ok) {
+      error_message = std::string{skus::result_to_string(result)};
+    }
+
+    std::move(callback_state->cb).Run(error_message);
```
Comment (bridiver) [inline components/skus/browser/skus_service_impl.cc:40]: Once chromium adds support for `base::expected` in mojo we will probably want to use that here instead. Personally I would change the JS api to return an object that has both a status and an error message and pass two params here instead of using a non-empty string as an error condition. Maybe for a follow-up? In general using a non-empty string to indicate an error is highly discouraged

Comment (bridiver) [reply components/skus/browser/skus_service_impl.cc:40]: The empty/non-empty string is especially problematic because some code uses it the opposite way: an empty string indicates an error. It's better to separate the result from any error message text.

File: components/skus/renderer/skus_js_handler.cc [hunk 5]
```diff
@@ -194,10 +194,13 @@ void SkusJSHandler::OnFetchOrderCredentials(
...
-  v8::Local<v8::String> result;
-  result = v8::String::NewFromUtf8(isolate, response.c_str()).ToLocalChecked();
 
-  std::ignore = resolver->Resolve(context, result);
+  // Create an empty string as the result
+  v8::Local<v8::String> empty_result =
+      v8::String::NewFromUtf8(isolate, "").ToLocalChecked();
+
+  // Always resolve the promise with the empty string
+  std::ignore = resolver->Resolve(context, empty_result);
```
Comment (nvonpentz) [inline components/skus/renderer/skus_js_handler.cc:203]: Temporarily making this always resolve to the empty string in order to make it easier to debug why the promise does not resolve.

File: renderer/skus/js_skus_browsertest.cc [hunk 6]
```diff
@@ -52,4 +55,28 @@ TEST_F(JsSkusBrowserTest, AttachSkus) {
   EXPECT_TRUE(ExecuteJavascript(command));
 }
 
+TEST_F(JsSkusBrowserTest, FetchOrderCredentials) {
+  SkusRenderFrameObserver observer(GetMainRenderFrame());
+  GURL url("https://account.brave.software");
+  LoadHTMLWithUrlOverride(R"(<html><body></body></html>)", url.spec().c_str());
+
+  blink::WebScriptSource source = blink::WebScriptSource(
+      // blink::WebString::FromASCII("new Promise((resolve, reject) => { resolve('hi'); });"));
```
Comment (nvonpentz) [inline renderer/skus/js_skus_browsertest.cc:64]: This line works -- the code in this test will await the promise.

File: renderer/skus/js_skus_browsertest.cc [hunk 7]
```diff
@@ -52,4 +55,28 @@ TEST_F(JsSkusBrowserTest, AttachSkus) {
...
+      blink::WebString::FromASCII("window.chrome.braveSkus.fetch_order_credentials('', '');"));
```
Comment (nvonpentz) [inline renderer/skus/js_skus_browsertest.cc:65]: This code does not work. The promise is not awaited.

File: components/skus/browser/skus_service_impl.cc (hunk 4 above)
Comment (bridiver) [reply components/skus/browser/skus_service_impl.cc:40]: also if you define skus::SkusResult in mojom you can pass it directly to the callback

File: components/skus/renderer/skus_utils.cc [hunk 8]
```diff
@@ -23,8 +23,14 @@ bool IsSafeOrigin(const blink::WebSecurityOrigin& origin) {
        {blink::WebSecurityOrigin::Create(
            GURL("https://account.brave.software"))}}};
 
+  const GURL origin_url(origin.ToString().Utf8());
```
[the remark this answers is not in the document]
Comment (fmarier) [reply components/skus/renderer/skus_utils.cc:26]: This specific alert is a false positive because a security origin is created further down.

That said, what's the purpose of removing the port from the origin? Could a comment be added here?

File: components/skus/renderer/skus_utils.cc [hunk 9]
```diff
@@ -23,8 +23,14 @@ bool IsSafeOrigin(const blink::WebSecurityOrigin& origin) {
...
+  const GURL stripped_port_url =
+      GURL(origin_url.scheme() + "://" + origin_url.host() + origin_url.path());
+  blink::WebSecurityOrigin stripped_port_origin =
+      blink::WebSecurityOrigin::Create(stripped_port_url);
+
   for (const blink::WebSecurityOrigin& safe_origin : *safe_origins) {
-    if (safe_origin.IsSameOriginWith(origin)) {
+    if (safe_origin.IsSameOriginWith(stripped_port_origin)) {
       return true;
     }
```
Comment (supermassive) [inline components/skus/renderer/skus_utils.cc:35]: Code for test-only purposes should not affect production behavior. To handle random port of embedded server use something like this in SetUpCommandLine:

```    
command_line->AppendSwitchASCII(
        network::switches::kHostResolverRules,
        base::StringPrintf("MAP account.brave.software:443 127.0.0.1:%d",
                           https_server_.port()));
```
Or search for `kHostResolverRules` for more examples.

File: renderer/skus/BUILD.gn [hunk 10]
```diff
@@ -8,7 +8,10 @@ import("//testing/test.gni")
 source_set("browser_tests") {
   testonly = true
 
-  sources = [ "js_skus_browsertest.cc" ]
+  sources = [
+    "js_skus_browsertest.cc",
+    "js_skus_platform_browsertest.cc",
+  ]
```
Comment (supermassive) [inline renderer/skus/BUILD.gn:15]: `deps` section should be fixed so `gn_check` is happy. Otherwise it may compile incorrectly and don't work at runtime as expected.

File: renderer/skus/js_skus_platform_browsertest.cc [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+std::unique_ptr<net::test_server::HttpResponse> HandleRequest(
+    const net::test_server::HttpRequest& request) {
+  auto response = std::make_unique<net::test_server::BasicHttpResponse>();
+  response->set_code(net::HTTP_OK);
+  response->set_content(R"({})");
+  response->set_content_type("application/json");
+  return response;
+}
+
```
Comment (supermassive) [inline renderer/skus/js_skus_platform_browsertest.cc:46]: wrap in anonymous namespace please

File: renderer/skus/js_skus_platform_browsertest.cc [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+class JsSkusPlatformBrowserTest : public PlatformBrowserTest {
+ public:
+  JsSkusPlatformBrowserTest()
+      : https_server_(net::EmbeddedTestServer::TYPE_HTTPS) {
+    scoped_feature_list_.InitWithFeatures({skus::features::kSkusFeature}, {});
+  }
+
```
Comment (supermassive) [inline renderer/skus/js_skus_platform_browsertest.cc:53]: Inline construction is preferred:
```
net::EmbeddedTestServer https_server_{net::EmbeddedTestServer::TYPE_HTTPS};
```
and
```
base::test::ScopedFeatureList scoped_feature_list_{skus::features::kSkusFeature};
```

File: components/skus/renderer/skus_utils.cc (hunk 9 above)
Comment (nvonpentz) [reply components/skus/renderer/skus_utils.cc:35]: I'm not sure this will work in this case. If I understand right, the port of the`origin` parameter in IsSafeOrigin is still the randomly assigned value, even if we route traffic to account.brave.software:443 to the randomly generated port. I tried starting the https_server_ at 443, but I get a permissions denied error.

File: renderer/skus/BUILD.gn [hunk 13]
```diff
@@ -8,15 +8,24 @@ import("//testing/test.gni")
 source_set("browser_tests") {
   testonly = true
 
-  sources = [ "js_skus_browsertest.cc" ]
+  sources = [
```
Comment (bridiver) [inline renderer/skus/BUILD.gn:11]: these should be moved into browser/skus now since we're not using the renderer test

File: renderer/skus/BUILD.gn [hunk 14]
```diff
@@ -8,15 +8,24 @@ import("//testing/test.gni")
...
+  ]
 
   defines = [ "HAS_OUT_OF_PROC_TEST_RUNNER" ]
 
   deps = [
+    "//brave/components/script_injector/common/mojom",
+    "//brave/components/skus/browser:browser",
     "//brave/components/skus/common",
     "//brave/components/skus/renderer",
+    "//chrome/browser",
```
[the remark this answers is not in the document]
Comment (bridiver) [reply renderer/skus/BUILD.gn:23]: also this should be `//chrome/browser:browser_process`

File: renderer/skus/BUILD.gn [hunk 15]
```diff
@@ -8,15 +8,24 @@ import("//testing/test.gni")
...
     "//chrome/common",
+    "//chrome/test:test_support",
+    "//chrome/test:test_support_ui",
     "//content/test:test_support",
+    "//net:test_support",
```
Comment (bridiver) [inline renderer/skus/BUILD.gn:28]: something also seems wrong here because this should need
```
if (is_android) {
    deps += [ "//chrome/test:test_support_ui_android" ]
  } else {
    deps += [ "//chrome/test:test_support_ui" ]
  }
```
for PlatformBrowserTest

Comment (bridiver) [reply renderer/skus/BUILD.gn:28]: it's also missing deps for components/prefs, etc... so something is wrong with gn check here

File: browser/skus/BUILD.gn [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+  deps = [
+    "//brave/components/script_injector/common/mojom",
+    "//brave/components/skus/browser",
+    "//brave/components/skus/common",
+    "//chrome/browser",
+    "//chrome/browser:browser_process",
+    "//chrome/browser/ui",
+    "//chrome/test:test_support",
+    "//chrome/test:test_support_ui",
```
Comment (bridiver) [inline browser/skus/BUILD.gn:20]: you have this listed twice

File: browser/skus/BUILD.gn [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+  sources = [ "js_skus_api_browsertest.cc" ]
+
+  deps = [
+    "//brave/components/script_injector/common/mojom",
+    "//brave/components/skus/browser",
+    "//brave/components/skus/common",
+    "//chrome/browser",
+    "//chrome/browser:browser_process",
+    "//chrome/browser/ui",
```
Comment (bridiver) [inline browser/skus/BUILD.gn:18]: I don't see this used in the test

File: browser/skus/BUILD.gn [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+    "//chrome/test:test_support",
+    "//chrome/test:test_support_ui",
+    "//content/test:test_support",
```
Comment (bridiver) [inline browser/skus/BUILD.gn:21]: you're still missing deps here for net/test, etc... and you have some extra ones you don't need.. did you run gn check?

File: browser/skus/js_skus_api_browsertest.cc [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+#include "brave/components/skus/browser/pref_names.h"
+#include "brave/components/skus/common/features.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/test/base/chrome_test_utils.h"
+#include "components/prefs/pref_service.h"
+#include "content/public/common/content_client.h"
+#include "content/public/test/browser_test.h"
+#include "content/public/test/browser_test_utils.h"
+#include "content/public/test/content_mock_cert_verifier.h"
+#include "mojo/public/cpp/bindings/associated_remote.h"
```
Comment (bridiver) [inline browser/skus/js_skus_api_browsertest.cc:20]: this include does not appear to be used

File: browser/skus/js_skus_api_browsertest.cc [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+
+#include "base/test/scoped_feature_list.h"
+#include "brave/browser/brave_content_browser_client.h"
+#include "brave/components/script_injector/common/mojom/script_injector.mojom.h"
+#include "brave/components/skus/browser/pref_names.h"
+#include "brave/components/skus/common/features.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/test/base/chrome_test_utils.h"
+#include "components/prefs/pref_service.h"
+#include "content/public/common/content_client.h"
```
Comment (bridiver) [inline browser/skus/js_skus_api_browsertest.cc:16]: I don't think this is needed anymore either. Please double check to make sure you have all the includes you need and no extras

File: browser/skus/BUILD.gn [hunk 21]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2023 The Brave Authors. All rights reserved.
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at https://mozilla.org/MPL/2.0/.
+
+source_set("browser_tests") {
```
Comment (bridiver) [inline browser/skus/BUILD.gn:6]: actually please move this BUILD.gn and the browsertest to browser/skus/test so we don't have a BUILD.gn and sources.gni in the same directory. Sorry, I didn't see what was in this before

**Review by bridiver — APPROVED** [review]
